### ISSUE TITLE
docker service ls reports all replicas are up even when a worker node is down

### ISSUE THREAD

**Issue by kavinduchamiran** [moby/moby#45922]
### Description

In a multi-node docker swarm cluster, when a worker node is offline, the docker service ls command's output shows the replica count of the services that were assigned to that node as 1/1 (actual = expected). The service was only assigned to the node that went offline and a replacement container is **not** running on any other node.

### Reproduce

1. Create a multi-node docker swarm cluster.
2. Deploy a few Docker swarm services using docker stack deploy. Don't assign the same service to more than one node.
3. Get the output of the docker node ls command. It should show that all nodes are online.
4. Get the output of the docker service ls command. It should show the replica count of each service to be (actual == expected)
5. Shut down one of the worker nodes.
6. Get the output of the docker node ls command. It should show that one of the nodes is offline now.
7. Get the output of the docker service ls command again. It should still show the replica count of each service to be (actual == expected)

### Expected behavior

docker service ls output should show that the replica count of the services that were assigned to the offline node as 0/expected.

### docker version

```bash
Client: Docker Engine - Community
 Version:           20.10.22
 API version:       1.41
 Go version:        go1.18.9
 Git commit:        3a2c30b
 Built:             Thu Dec 15 22:28:04 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.22
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.9
  Git commit:       42c8b31
  Built:            Thu Dec 15 22:25:49 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.15
  GitCommit:        5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  compose: Docker Compose (Docker Inc., v2.14.1)
  scan: Docker Scan (Docker Inc., v0.23.0)

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 2
 Server Version: 20.10.22
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: khrw23o2eum1o3r23pffaqugv
  Is Manager: true
  ClusterID: x11cdrphzkknzlds6oh42w46d
  Managers: 1
  Nodes: 4
  Default Address Pool: 10.10.0.0/20  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.0.0.4
  Manager Addresses:
   10.0.0.4:2377
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.0-1030-aws
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 3.773GiB
 Name: ip-10-0-0-4
 ID: RFCW:TLXB:DJBW:IBUA:3EJK:IEXR:MI7D:KS7B:HTJ5:BUTU:YZD6:6Q66
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
  private_ip=10.0.0.4
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"

**Comment by devchris123** [comment]
Can you show the stack file you used to deploy the service?

I tried to reproduce this behaviour (in order to test it if I can fix it) with an example web service.
In my case, once I shut down the second worker, (the only worker running the web service), the manager node took over and the service replicas I got output from docker service ls was 2/1.
Maybe in your case you configured your manager node not to run additional containers if containers on worker nodes fail?

My setup was:
2 VMs with Ubuntu 20.04.2
Hosts: Worker1 (manager node) Worker2 (worker node)
Worker1:
Docker daemon version 20.10 (compiled binaries from branch 20.10 in a dev container and copied to Worker1)
Worker2:
Docker daemon version 24.0.4
Services:
I used the example service from https://docs.docker.com/engine/swarm/stack-deploy/
stack file:
```
version: "3"

services:

  web:

    image: 127.0.0.1:5000/stackdemo

    build: .

    ports:

      - "8000:8000"

    environment:

      - "constraint:node==Worker2"

  redis:

    image: redis:alpine

    environment:

      - "constraint:node==Worker"
```

Relevant outputs before taking Worker2 down:

docker node ls
```
ID                            HOSTNAME   STATUS    AVAILABILITY   MANAGER STATUS   ENGINE VERSION

pc4be1uvna7940qgmr7s8sqic *   Worker1    Ready     Active         Leader           dev

cei0aqzjlduy8ljnvmu3fpo1u     Worker2    Ready     Active                          24.0.4
````

docker service ls
```
ID             NAME              MODE         REPLICAS   IMAGE                             PORTS

eqd3p5cx95bu   registry          replicated   1/1        registry:2                        *:5000->5000/tcp

jph9ntk9l7jy   stackdemo_redis   replicated   1/1        redis:alpine                      

fy00wpidzmip   stackdemo_web     replicated   1/1        127.0.0.1:5000/stackdemo:latest   *:8000->8000/tcp
```

docker stack ps stackdemo
```
ID             NAME                IMAGE                             NODE      DESIRED STATE   CURRENT STATE                    ERROR     PORTS

kxzmn8bcc1zy   stackdemo_redis.1   redis:alpine                      Worker1   Running         Running less than a second ago             

laciiga404dp   stackdemo_web.1     127.0.0.1:5000/stackdemo:latest   Worker2   Running         Running 4 seconds ago
```

Relevant outputs after taking Worker2 down
docker node ls
```
ID                            HOSTNAME   STATUS    AVAILABILITY   MANAGER STATUS   ENGINE VERSION

pc4be1uvna7940qgmr7s8sqic *   Worker1    Ready     Active         Leader           dev

cei0aqzjlduy8ljnvmu3fpo1u     Worker2    Down      Active                          24.0.4
```
docker service ls
```
ID             NAME              MODE         REPLICAS   IMAGE                             PORTS

eqd3p5cx95bu   registry          replicated   1/1        registry:2                        *:5000->5000/tcp

jph9ntk9l7jy   stackdemo_redis   replicated   1/1        redis:alpine                      

fy00wpidzmip   stackdemo_web     replicated   2/1        127.0.0.1:5000/stackdemo:latest   *:8000->8000/tcp
```
docker stack ps stackdemo
```
ID             NAME                  IMAGE                             NODE      DESIRED STATE   CURRENT STATE                ERROR     PORTS

kxzmn8bcc1zy   stackdemo_redis.1     redis:alpine                      Worker1   Running         Running 3 minutes ago                  

ftues2qwhpze   stackdemo_web.1       127.0.0.1:5000/stackdemo:latest   Worker1   Running         Running about a minute ago             

laciiga404dp    \_ stackdemo_web.1   127.0.0.1:5000/stackdemo:latest   Worker2   Shutdown        Running 3 minutes ago    
```

**Comment by kavinduchamiran** [comment]
@devchris123 Thanks for the information. 

Yes, my manager node does **not** run any containers.

However, if Docker is not updating the service status when a node goes down, isn't that displaying obsolete or incorrect information to the `docker service ls` command?

Is there a way to force-trigger a service status update? We are using the output of the command to extract services and replicas to be shown in a UI dashboard we developed. And due to this error, it is now showing false data.

**Comment by devchris123** [comment]
@kavinduchamiran 
From docker swarms manager perspective, the manager notices, that the node is unresponsive. It does update the desired status of the task running the service, but it will not update the current status of the task, which only the node itself knows. The manager will update the task from running -> shutdown as soon as the node comes up and follows the managers instructions, i.e. shutdown the task as desired and report current shutdown status.

I opened a PR to "fix" this: 
https://github.com/moby/swarmkit/pull/3146
It was suggested, that it might be better to fix this on the client, since the manager shouldn't assume too much about the other nodes, it was argued ( for all it knows the task could still be running somewhere ). Although depending on the configuration, the manager is pessimistic about inactive nodes and will spawn new tasks replacing the olds ones (but not updating their current statuses, thus the confusing output).

I found that in an older implementation the client calculated the running tasks for the service on all nodes and exluded the tasks on inactive nodes: 
You can find it right here: https://github.com/docker/cli/blob/cdabfa2aa55a3560d0fe5f63074571afae840ab3/cli/command/service/list.go in line 167.
In a newer API version the daemon does the calculations, and does not exclude running tasks on inactive nodes.

Long story short: I'll try to make a change to the client and open an PR. But it'd be nice if someone more experienced with the codebase would suggest, where the change would fit best.

If you'd query the daemon or have more information about node status, you could put together your own interpretation of running tasks. Or add some information about running tasks on inactive nodes.
